File: Windows Phone Starter Kits/APIMASH_CNorris_StarterKit/APIMASHLib/APIMASHLib/APIMASHMap.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Runtime.Serialization.Json;
using System.Text;
using System.Threading.Tasks;
using System.Xml.Serialization;
using Newtonsoft.Json;

/*
* LICENSE: https://raw.github.com/apimash/StarterKits/master/LicenseTerms-SampleApps%20.txt
*/

/*
 *
 *  A P I   M A S H 
 *
 * This class makes the HTTP call and deserialzies the stream to a supplied Type
*/

namespace APIMASHLib
{
    public static class APIMASHMap
    {
#if OBJMAP_JSON // .NET JSON
        public static string SerializeObject(object objectToSerialize)
        {
            using (var stream = new MemoryStream())
            {
                try
                {
                    var serializer = new DataContractJsonSerializer(objectToSerialize.GetType());
                    serializer.WriteObject(stream, objectToSerialize);
                    stream.Position = 0;
                    var reader = new StreamReader(stream);
                    return reader.ReadToEnd();
                }
                catch (Exception e)
                {
                    Debug.WriteLine("Serialize:" + e.Message);
                    return string.Empty;
                }
            }
        }

        public static T DeserializeObject<T>(string objString)
        {
            using (var stream = new MemoryStream(Encoding.Unicode.GetBytes(objString)))
            {
                try
                {
                    var serializer = new DataContractJsonSerializer(typeof(T));
                    return (T)serializer.ReadObject(stream);
                }
                catch (Exception) { throw; }
            }
        }
#endif

#if OBJMAP_XML // .NET XML
        public static T DeserializeObject<T>(string objString)
        {
            using (MemoryStream _Stream = new MemoryStream(Encoding.Unicode.GetBytes(objString)))
            {
                try
                {

                    DataContractJsonSerializer _serializer = new DataContractJsonSerializer(typeof(T));
                    return (T)_serializer.ReadObject(_Stream);

                    XmlSerializer _serializer = new XmlSerializer(typeof(T));
                    return (T)_serializer.Deserialize(_Stream);
                }
                catch (Exception) { throw; }
            }
        }
#endif

#if OBJMAP_NEWTONSOFT // JSON.NET
        public static T DeserializeObject<T>(string objString)
        {
            var settings = new JsonSerializerSettings {NullValueHandling = NullValueHandling.Ignore};
            return (T)JsonConvert.DeserializeObject<T>(objString, settings);
        }
#endif

        private static string DecompressBytes(byte[] compressedBytes)
        {
            string uncompressedObj = string.Empty;
            // http://stackoverflow.com/questions/12894406/inflating-a-compressed-byte-array-in-winrt
            // The reason for skipping the first two bytes is that they are part of the zlib spec and not the deflate spec: 
            // http://george.chiramattel.com/blog/2007/09/deflatestream-block-length-does-not-match.html
            using (System.IO.Compression.GZipStream stream = new System.IO.Compression.GZipStream(new MemoryStream(compressedBytes, 0, compressedBytes.Length), System.IO.Compression.CompressionMode.Decompress))
            {
                MemoryStream memory = new MemoryStream();
                byte[] writeData = new byte[4096];
                int resLen;
                while ((resLen = stream.Read(writeData, 0, writeData.Length)) > 0)
                {
                    memory.Write(writeData, 0, resLen);
                }
                var uncompressedBytes = memory.ToArray();
                uncompressedObj = Encoding.UTF8.GetString(uncompressedBytes, 0, uncompressedBytes.Length);
            }
            return uncompressedObj;
        }

        async public static Task<T> LoadObject<T>(string apiCall)
        {
            var ws = new HttpClient();
            var uriAPICall = new Uri(apiCall);
            var objString = await ws.GetStringAsync(uriAPICall);
            return (T)DeserializeObject<T>(objString);
        }

        async public static Task<T> LoadCompressedObject<T>(string apiCall)
        {
            var ws = new HttpClient();
            ws.DefaultRequestHeaders.AcceptEncoding.Add( StringWithQualityHeaderValue.Parse("DEFLATE") );
            var uriAPICall = new Uri(apiCall);
            var objArray = await ws.GetByteArrayAsync(uriAPICall);
            var uncompressedString = DecompressBytes(objArray);
            return (T)DeserializeObject<T>(uncompressedString);
        }
    }
}

File: Windows Starter Kits/APIMASH_CNorris_VB_StarterKit/APIMASHVBLib/APIMashMap.vb
Imports System.Diagnostics
Imports System.IO
Imports System.Net
Imports System.Net.Http
Imports System.Net.Http.Headers
Imports System.Runtime.Serialization.Json
Imports System.Text
Imports System.Threading.Tasks
Imports System.Xml.Serialization
Imports Newtonsoft.Json

'
'* LICENSE: https://raw.github.com/apimash/StarterKits/master/LicenseTerms-SampleApps%20.txt
'

'
' *
' *  A P I   M A S H 
' *
' * This class makes the HTTP call and deserialzies the stream to a supplied Type
'

Namespace APIMASHLib
    Public NotInheritable Class APIMASHMap
        Private Sub New()
        End Sub
#If OBJMAP_JSON Then
		Public Shared Function SerializeObject(objectToSerialize As Object) As String
			Using stream = New MemoryStream()
				Try
					Dim serializer = New DataContractJsonSerializer(objectToSerialize.[GetType]())
					serializer.WriteObject(stream, objectToSerialize)
					stream.Position = 0
					Dim reader = New StreamReader(stream)
					Return reader.ReadToEnd()
				Catch e As Exception
					Debug.WriteLine("Serialize:" & e.Message)
					Return String.Empty
				End Try
			End Using
		End Function

		Public Shared Function DeserializeObject(Of T)(objString As String) As T
			Using stream = New MemoryStream(Encoding.Unicode.GetBytes(objString))
				Try
					Dim serializer = New DataContractJsonSerializer(GetType(T))
					Return DirectCast(serializer.ReadObject(stream), T)
				Catch generatedExceptionName As Exception
					Throw
				End Try
			End Using
		End Function
#End If

#If OBJMAP_XML Then
		Public Shared Function DeserializeObject(Of T)(objString As String) As T
			Using _Stream As New MemoryStream(Encoding.Unicode.GetBytes(objString))
				Try

					Dim _serializer As New DataContractJsonSerializer(GetType(T))
					Return DirectCast(_serializer.ReadObject(_Stream), T)

					Dim _serializer As New XmlSerializer(GetType(T))
					Return DirectCast(_serializer.Deserialize(_Stream), T)
				Catch generatedExceptionName As Exception
					Throw
				End Try
			End Using
		End Function
#End If

#If OBJMAP_NEWTONSOFT Then
		Public Shared Function DeserializeObject(Of T)(objString As String) As T
            Dim settings = New JsonSerializerSettings() With { _
                 .NullValueHandling = NullValueHandling.Ignore _
            }
			Return DirectCast(JsonConvert.DeserializeObject(Of T)(objString, settings), T)
		End Function
#End If

        Private Shared Function DecompressBytes(compressedBytes As Byte()) As String
            Dim uncompressedObj As String = String.Empty
            ' http://stackoverflow.com/questions/12894406/inflating-a-compressed-byte-array-in-winrt
            ' The reason for skipping the first two bytes is that they are part of the zlib spec and not the deflate spec: 
            ' http://george.chiramattel.com/blog/2007/09/deflatestream-block-length-does-not-match.html
            Using stream As New System.IO.Compression.GZipStream(New MemoryStream(compressedBytes, 0, compressedBytes.Length), System.IO.Compression.CompressionMode.Decompress)
                Dim memory As New MemoryStream()
                Dim writeData As Byte() = New Byte(4095) {}
                Dim resLen As Integer
                While (InlineAssignHelper(resLen, stream.Read(writeData, 0, writeData.Length))) > 0
                    memory.Write(writeData, 0, resLen)
                End While
                Dim uncompressedBytes = memory.ToArray()
                uncompressedObj = Encoding.UTF8.GetString(uncompressedBytes, 0, uncompressedBytes.Length)
            End Using
            Return uncompressedObj
        End Function

        Public Shared Async Function LoadObject(Of T)(apiCall As String) As Task(Of T)
            Dim ws = New HttpClient()
            Dim uriAPICall = New Uri(apiCall)
            Dim objString = Await ws.GetStringAsync(uriAPICall)
            Return DirectCast(DeserializeObject(Of T)(objString), T)
        End Function

        Public Shared Async Function LoadCompressedObject(Of T)(apiCall As String) As Task(Of T)
            Dim ws = New HttpClient()
            ws.DefaultRequestHeaders.AcceptEncoding.Add(StringWithQualityHeaderValue.Parse("DEFLATE"))
            Dim uriAPICall = New Uri(apiCall)
            Dim objArray = Await ws.GetByteArrayAsync(uriAPICall)
            Dim uncompressedString = DecompressBytes(objArray)
            Return DirectCast(DeserializeObject(Of T)(uncompressedString), T)
        End Function
        Private Shared Function InlineAssignHelper(Of T)(ByRef target As T, value As T) As T
            target = value
            Return value
        End Function
    End Class
End Namespace
